Add opt for gpuarray.GpuElemwise so exponent of pow has same dtype as output

carriepl opened the pull request:
In the gpuarray backend, pow can only be done on inputs that are both float (producing a float output) or both double (producing a double output). Currently, some unit tests for Scan attempt to do `a_float ** an_int16` which raises a compilation error. This optimization adds  a cast to ensure that the exponent has the same dtype as the output to prevent this.

nouiz commented on theano/sandbox/gpuarray/opt.py:
```diff
@@ -260,6 +261,25 @@ def local_gpu_elemwise(node):
 
+@register_opt()
+@gof.local_optimizer([GpuElemwise])
+def cast_elemwise_pow_exponent(node):
```
Why this is needed? This also need a test.

carriepl replied:
I just updated the description. I just wanted to create the PR as soon as possible so Travis would start faster. Let me know if anything is still needed

abergeron replied:
It would still need a test.

carriepl replied:
I added tests. Now, if the base is float32 or float64, the exponent can have any uint, int or float dtype and it works. However, there are still cases where it doesn't work. If both the base and the exponent are ints, for instance, the output dtype will also be int and it will not fit in the two patterns that the gpu implementation supports at this point in time (float32,float32)->(float32) and (float64,float64)->(float64).

I am not sure which way would be best to handle this.
We could add casts before the inputs and after the output when moving the Elemwise(pow) to the Gpu but that means in some cases we would have to downcast the output (say from float32 to int8). 
We could also not move the Elemwise(pow) nodes with unsupported dtypes to gpu but if the goal of the new backend was to support other dtypes than floating points, then that's probably not the best solution.
Any opinion on this?

carriepl replied:
@lamblin, what do you think?

lamblin commented:
Summary of IRL discussion:
- Both numpy and the current C code for `power` seem to cast their arguments to float or double, then do the exponentiation and cast back, all of that implicitly. One way of exposing that is to compute `(-1) ** x` for a large odd `x` (for instance, `2**60 - 1`)
- So, explicitly doing these conversion on the GPU makes sense. We could do that by changing the C code, or adding casts in the graph, I think adding casts in the graph would be easier.
- If we want to make things better for integer-valued exponentiation, I think we should not use the `power` op when `**` is called on integers, but have a different Op for that. For instance, Python gives the right result if both sides of the exponent are integers, but rounds things if one of the arguments is a float:
```python
>>> (-1) ** (2**60 + 1)
-1
>>> (-1.) ** (2**60 + 1)
1.0
```

carriepl commented:
Following the discussion with @lamblin, I implemented the solution in which casts are added to the graph, if needed. I'm not 100% certain the way I did the casts on the GPU is the simplest/cleanest way  so if there's anything to improve there, don't hesitate.

lamblin commented on theano/sandbox/gpuarray/opt.py:
```diff
@@ -254,10 +255,43 @@ def local_gpu_elemwise(node):
-                      nfunc_spec=op.nfunc_spec)
-    return res
+                          inplace_pattern=copy.copy(op.inplace_pattern),
+                          nfunc_spec=op.nfunc_spec)
```
Strange indentation here

carriepl replied:
Done.

lamblin commented on theano/sandbox/gpuarray/opt.py:
```diff
@@ -254,10 +255,43 @@ def local_gpu_elemwise(node):
+    if isinstance(op.scalar_op, Pow):
+        old_out_dtype = node.outputs[0].dtype
+        old_inp_dtypes = [inp.dtype for inp in node.inputs]
+        new_out_dtype = upcast("float32", *old_inp_dtypes)
```
It makes sense to me, I'm just wondering if we need a special case for float16.

carriepl replied:
I'm not sure. @abergeron, what do you think?

abergeron replied:
I think you should just upcast all the argument to `upcast(*old_inp_dtypes)` and not include a "float32".

This way it should work for float16.

carriepl replied:
The upcast was meant to ensure that the resulting dtype would be floating point and not an integer (which is not currently supported). Can't I simply upcast with 'float16' instead of 'float32'?

abergeron replied:
Yes that would work.

lamblin replied:
Or maybe use floatX, unless floatX is float64, in which case use float32.
I'm afraid that int8 ** int8 gets casted to float16 even in cases where float16 is not supported.

abergeron replied:
float16 is always supported for Elemwise operations on the GPU.  It might not be optimal, but it will work.  And since we cast the result back to the previous type, it won't make much of a difference.

However using floatX is fine for me.

carriepl replied:
It doesn't appear to be always supported. Without the float16 kernels installed (since they're still an optional dependency), the compilation raises an error (described in the other comment thread below).

abergeron commented on theano/sandbox/gpuarray/tests/test_elemwise.py:
```diff
@@ -46,6 +46,26 @@ def test_c_inplace(self):
+
+    dtypes = ["uint8", "uint16", "uint32", "uint64",
+              "int8", "int16", "int32", "int64",
+              "float32", "float64"]
```
Perhaps you can add "float16" to that list.

carriepl replied:
Adding "float16" makes compilation crash without the kernels for float16 installed (the same error as with integer types before : "calling std::pow from a kernel function is forbidden and you shouldn't") so I'm not certain it's a good idea.

nouiz commented:
This need a rebase.

carriepl commented:
Added 'float16' to the list of dtypes to test and did the requested rebase.
Upcasting with 'float16' instead of 'float32' cannot be done for now because of a bug identified in GpuElemwise by @abergeron. I added a comment to document this.

abergeron commented:
This is ok (after travis passes).  I've made https://github.com/Theano/Theano/issues/3138 to keep track of the GpuElemwise bug.

nouiz commented on theano/sandbox/gpuarray/opt.py:
```diff
@@ -262,10 +263,49 @@ def local_gpu_elemwise(node):
+                gpu_cast_op = GpuElemwise(Cast(Scalar(new_out_dtype)))
+                new_inputs.append(gpu_cast_op(gpu_from_host(inp)))
+            else:
+                new_inputs.append(gpu_from_host(inp))
```
use as_gpuarray_variable not, gpu_from_host.

nouiz commented on theano/sandbox/gpuarray/tests/test_elemwise.py:
```diff
@@ -46,6 +46,26 @@ def test_c_inplace(self):
+            base = theano.tensor.vector(dtype=dtype_base)
+            exp = theano.tensor.vector(dtype=dtype_exp)
+            output = base ** exp
+            f = theano.function([base, exp], output)
```
use the function to make sure you have the right answer.

carriepl commented:
Done. Should be all fine now.
